### DISCUSSION TITLE
with_polymorphic and aliasing

### DISCUSSION THREAD

**Discussion by sonthonaxrk** [sqlalchemy/sqlalchemy#7797] [Usage Questions]
So really difficult question about with_polymorphic and mapping back to ORM objects.

I've been doing some query optimisation, and annoyingly I have to refactor a union query from this:

```sql
SELECT metadata.data, pjoin.id FROM (
   SELECT * FROM a
   UNION ALL
   SELECT * FROM b
) pjoin
JOIN metadata pjoin.id == metadata.id
```

To this, because there query planner doesn't pick up on the distributive nature of the join onto metadata.  They're functionally equivalent, but the former causes sequential scans on `a` and `b`. 

```sql
SELECT pjoin.metadata_data, pjoin.id FROM (
   SELECT id FROM a
   JOIN metadata a.id == metadata.id
   UNION ALL
   SELECT id FROM b
   JOIN metadata b.id == metadata.id
) pjoin
```

However, I'm having a lot of difficulty mapping this back to the ORM.  Here's my attempt:

```python
import pytest
from sqlalchemy import *
from sqlalchemy.orm import *
from sqlalchemy.ext.declarative import (declarative_base, declared_attr,
    AbstractConcreteBase)
from textwrap import dedent
from sqlalchemy import *
from sqlalchemy.orm import aliased, with_polymorphic
from sqlalchemy.orm.util import AliasedClass
from sqlalchemy.sql import coercions, roles
from sqlalchemy.sql import util as sql_util
from sqlalchemy import exc as sa_exc
from sqlalchemy import sql, util
from collections import OrderedDict
from sqlalchemy.sql.expression import LABEL_STYLE_DISAMBIGUATE_ONLY
from sqlalchemy.sql.expression import LABEL_STYLE_TABLENAME_PLUS_COL


Base = declarative_base()

class Metadata(Base):
    __tablename__ = "metadata"
    id = Column(
        Integer,
        primary_key=True,
        unique=True,
    )


class BaseObj(AbstractConcreteBase, Base):
    @declared_attr
    def id(cls):
        return Column(ForeignKey(Metadata.id), primary_key=True)


class A(BaseObj):
    __tablename__ = "a"
    thing1 = Column(String)

    @declared_attr
    def __mapper_args__(cls):
        return {
            "polymorphic_identity": cls.__tablename__,
            "concrete": True
        }


class B(BaseObj):
    __tablename__ = "b"
    thing2 = Column(String)

    @declared_attr
    def __mapper_args__(cls):
        return {
            "polymorphic_identity": cls.__tablename__,
            "concrete": True
        }




e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

@pytest.fixture
def session():
    configure_mappers()
    yield Session(e)

# So the problem is moving the join into the union
# This is doable if I copy polymorphic_union and do some edits
# Pardon the copied code
def polymorphic_union(
        table_map,
        typecolname,
        aliasname="pjoin",
        cast_nulls=True,
        limit=100
):

    colnames = util.OrderedSet()
    colnamemaps = {}
    types = {}
    for key in table_map:
        table = table_map[key]

        table = coercions.expect(roles.StrictFromClauseRole, table, allow_select=True)
        table_map[key] = table

        m = {}
        for c in table.c:
            if c.key == typecolname:
                raise sa_exc.InvalidRequestError(
                    "Polymorphic union can't use '%s' as the discriminator "
                    "column due to mapped column %r; please apply the "
                    "'typecolname' "
                    "argument; this is available on "
                    "ConcreteBase as '_concrete_discriminator_name'" % (typecolname, c)
                )
            colnames.add(c.key)
            m[c.key] = c
            types[c.key] = c.type
        colnamemaps[table] = m

    def col(name, table):
        try:
            return colnamemaps[table][name]
        except KeyError:
            if cast_nulls:
                return sql.cast(sql.null(), types[name]).label(name)
            else:
                return sql.type_coerce(sql.null(), types[name]).label(name)

    result = []
    for type_, table in table_map.items():
        if typecolname is not None:

            result.append(
                (
                    sql.select(
                        Metadata,
                        *(
                            [col(name, table) for name in colnames]
                            + [
                                sql.literal_column(sql_util._quote_ddl_expr(type_)).label(
                                    typecolname
                                )
                            ]
                        )
                    )
                    .join(Metadata, Metadata.id == table.c.id)
                )
            )
        else:
            result.append(
                sql.select(*[col(name, table) for name in colnames]).select_from(table).limit(limit)
            )
    return sql.union_all(*result).alias(aliasname)


def test_query_joins_inside_union(session):
    special_union = polymorphic_union(
        dict((mp.polymorphic_identity, mp.local_table) for mp in [
            A.__mapper__,
            B.__mapper__,
        ]),
        "type",
    )

    base_obj_alias = with_polymorphic(
        BaseObj,
        [A, B],
        selectable=special_union,
        aliased=True,
    )

    # Succeeds in creating the following sql:
    print(select(base_obj_alias).from_statement(select(special_union)))
    # Which is good.  As it has everything we need
    """
    SELECT pjoin.id,
           pjoin.thing1,
           pjoin.id_1,
           pjoin.thing2,
           pjoin.type
    FROM
      (SELECT metadata.id AS id,
              a.thing1 AS thing1,
              a.id AS id_1,
              cast(NULL AS VARCHAR) AS thing2,
              'a' AS TYPE
       FROM a
       JOIN metadata ON metadata.id = a.id
       UNION ALL SELECT metadata.id AS id,
                        cast(NULL AS VARCHAR) AS thing1,
                        b.id AS id_2,
                        b.thing2 AS thing2,
                        'b' AS TYPE
       FROM b
       JOIN metadata ON metadata.id = b.id) AS pjoin
    """

    session.execute(
        select(base_obj_alias).from_statement(select(special_union))
    )
    # Raises
    # sqlalchemy.exc.NoSuchColumnError: "Could not locate column in row for column 'anon_1.id'"
```

**Comment by zzzeek** [comment]
The first thing I see is that the mapping looks exactly like a joined table inheritance mapping that's been broken out for some reason.   if you can get A and B to extend from MetaData, you can use joined inheritance which has much better querying patterns, where you can query simply the MetaData table and as it fetches results it can do a very fast IN query for the additional columns.

put another way, concrete inheritance with an abstract base should be avoided if at all possible.  any chance we can sidestep the whole problem that way here?

**Reply by zzzeek** [reply]
beyond that, basic use:

```py
  base_obj_alias = with_polymorphic(
        BaseObj,
        [A, B],
        selectable=special_union,
        aliased=True,
    )
    session.execute(
        select(base_obj_alias).from_statement(select(special_union))
    )
```

`base_obj_alias` is already against "special_union" so you would want to select it directly:

```
session.execute(select(base_obj_alias))
```

mixing that up with from_statement is likely getting into the weeds too much, i have no idea if that's expected to work so it seems like not.

**Reply by sonthonaxrk** [reply]
I'm afraid I can't really change the schema as that would be a big migration.  And another reason for using abstract base is that the union query is quite long, and I didn't want my response getting too wide.

On the other hand, the query performance is actually pretty good with the hand tuned SQL, so I'm really just fighting the ORM at this point.

Do you think I should just bite the bullet and load stuff manually?

I suppose I could make a simpler version of the same question: how do I map a AbstractBase to arbitrary SQL?

**Reply by zzzeek** [reply]
I wasnt suggesting to change the schema at all, just the model.   your schema is the joined table inheritance schema already.

**Reply by sonthonaxrk** [reply]
```
session.execute(select(base_obj_alias))
```

I'm afraid that doesn't work, since it produces 

```
SELECT pjoin_1.thing2,
       pjoin_1.id,
       pjoin_1.thing1,
       pjoin.type,
       pjoin_1.id_1
FROM
  (SELECT metadata.id AS id,
          a.thing1 AS thing1,
          a.id AS id_1,
          cast(NULL AS VARCHAR) AS thing2,
          'a' AS TYPE
   FROM a
   JOIN metadata ON metadata.id = a.id
   UNION ALL SELECT metadata.id AS id,
                    cast(NULL AS VARCHAR) AS thing1,
                    b.id AS id_2,
                    b.thing2 AS thing2,
                    'b' AS TYPE
   FROM b
   JOIN metadata ON metadata.id = b.id) AS pjoin_1,

  (SELECT b.thing2 AS thing2,
          b.id AS id,
          cast(NULL AS VARCHAR) AS thing1,
          'b' AS TYPE
   FROM b
   UNION ALL SELECT cast(NULL AS VARCHAR) AS thing2,
                    a.id AS id,
                    a.thing1 AS thing1,
                    'a' AS TYPE
   FROM a) AS pjoin%
```

Giving a cartesian join of the union.

**Reply by sonthonaxrk** [reply]
> your schema is the joined table inheritance schema already

Ah, I'll give it a think.  It's a lot of work.  Do you think that would be okay if I have a really wide table, like 250 rows across?  I might start to hit postgres limits.

**Reply by zzzeek** [reply]
OK aliased_class doesn't work here anyway because it can't locate the columns.  this is because the "abstract base" is mapped to an arbitrary selectable, and in order to make any kind of adaptation on that, the new "selectable" has to be based on the old one.

so the easiest way to get this right now is to use from_statement() and then make sure the selectable you are running is exactly in terms of the BaseObj table, this is most easy with text() where you'll get your exact SQL:

```py
def test_query_joins_inside_union(session):
    base_table = BaseObj.__mapper__.local_table

    result = session.scalars(
        select(BaseObj).from_statement(
            text(
                """
    SELECT pjoin.id,
           pjoin.thing1,
           pjoin.id_1,
           pjoin.thing2,
           pjoin.type
    FROM
      (SELECT metadata.id AS id,
              a.thing1 AS thing1,
              a.id AS id_1,
              cast(NULL AS VARCHAR) AS thing2,
              'a' AS TYPE
       FROM a
       JOIN metadata ON metadata.id = a.id
       UNION ALL SELECT metadata.id AS id,
                        cast(NULL AS VARCHAR) AS thing1,
                        b.id AS id_2,
                        b.thing2 AS thing2,
                        'b' AS TYPE
       FROM b
       JOIN metadata ON metadata.id = b.id) AS pjoin
"""
            ).columns(
                base_table.c.id,
                base_table.c.thing1,
                base_table.c.id.label("id_1"),
                base_table.c.thing2,
                base_table.c.type,
            )
        )
    )

    print(result.all())
```

POC with two rows of data follows (the part I wasnt sure about was the id.label("id_1") part)

```py
import pytest

from sqlalchemy import Column
from sqlalchemy import create_engine, text
from sqlalchemy import exc as sa_exc
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import select, union, null
from sqlalchemy import sql
from sqlalchemy import String
from sqlalchemy import util
from sqlalchemy.ext.declarative import AbstractConcreteBase
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.ext.declarative import declared_attr
from sqlalchemy.orm import configure_mappers
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import declared_attr
from sqlalchemy.orm import polymorphic_union
from sqlalchemy.orm import Session
from sqlalchemy.orm import with_polymorphic
from sqlalchemy.sql import coercions
from sqlalchemy.sql import roles
from sqlalchemy.sql import util as sql_util


Base = declarative_base()


class Metadata(Base):
    __tablename__ = "metadata"
    id = Column(
        Integer,
        primary_key=True,
        unique=True,
    )


class BaseObj(AbstractConcreteBase, Base):
    @declared_attr
    def id(cls):
        return Column(ForeignKey(Metadata.id), primary_key=True)

    def __repr__(self):
        return f"{self.__class__.__name__}({self.id})"

class A(BaseObj):
    __tablename__ = "a"
    thing1 = Column(String)

    @declared_attr
    def __mapper_args__(cls):
        return {"polymorphic_identity": cls.__tablename__, "concrete": True}


class B(BaseObj):
    __tablename__ = "b"
    thing2 = Column(String)

    @declared_attr
    def __mapper_args__(cls):
        return {"polymorphic_identity": cls.__tablename__, "concrete": True}


e = create_engine("sqlite://", echo='debug')
Base.metadata.create_all(e)

with Session(e) as sess:
    sess.add(Metadata(id=1))
    sess.add(Metadata(id=2))
    sess.flush()

    sess.add(A(id=1, thing1='thing1'))
    sess.add(B(id=2, thing2='thing2'))

    sess.commit()

@pytest.fixture
def session():
    configure_mappers()
    yield Session(e)


def test_query_joins_inside_union(session):
    base_table = BaseObj.__mapper__.local_table

    result = session.scalars(
        select(BaseObj).from_statement(
            text(
                """
    SELECT pjoin.id,
           pjoin.thing1,
           pjoin.id_1,
           pjoin.thing2,
           pjoin.type
    FROM
      (SELECT metadata.id AS id,
              a.thing1 AS thing1,
              a.id AS id_1,
              cast(NULL AS VARCHAR) AS thing2,
              'a' AS TYPE
       FROM a
       JOIN metadata ON metadata.id = a.id
       UNION ALL SELECT metadata.id AS id,
                        cast(NULL AS VARCHAR) AS thing1,
                        b.id AS id_2,
                        b.thing2 AS thing2,
                        'b' AS TYPE
       FROM b
       JOIN metadata ON metadata.id = b.id) AS pjoin
"""
            ).columns(
                base_table.c.id,
                base_table.c.thing1,
                base_table.c.id.label("id_1"),
                base_table.c.thing2,
                base_table.c.type,
            )
        )
    )

    print(result.all())
```

**Reply by zzzeek** [reply]
> > your schema is the joined table inheritance schema already
> 
> Ah, I'll give it a think. It's a lot of work. Do you think that would be okay if I have a really wide table, like 250 rows across? I might start to hit postgres limits.

is that a concern because...the JOIN that would happen between Metadata and a single class?  we dont have to have it joining all tables at once.   

So I can make a POC that does joined inheritance here, but the "load polymorphic on SELECTIN" feature I'm bragging about does not seem to support the case here where I need to manufacture the "polymorphic_on" part of it.      So...without a built-in "type" column on Metadata we have to use most of the same tricks anyway here, so probably not worth it, as the above concrete version keeps the "weird" part very local to that query you're doing which is overall safer.

**Reply by zzzeek** [reply]
guess I have to implement that, it's pretty wrong that it isnt working

**Reply by zzzeek** [reply]
OK it does actually work except for one tiny little bug I'm fixing in #7799

**Reply by sonthonaxrk** [reply]
> is that a concern because...the JOIN that would happen between Metadata and a single class?

So effectively I have a series of well defined objects (A, B), but their primary key maps to the primary key of common metadata.  And the primary API I've written is a polymorphic list of those objects.  Because the ordering is usually (but not always) by the common metadata id, I have no guarantee what types will be returned by the query.

The other reason that load polymorphic on SELECT in won't work in my case is that I might want to filter each thing in the UNION query.  Which is harder to do in in an optimal manner.

I'm under the impression the polymorphic join makes really wide queries, which is fine, but given I could get close to some arbitrary limit of table width, I'd rather use union.

Also zzzeek, thanks for your help.  Let me know if there's anything I can do for you.

**Reply by sonthonaxrk** [reply]
> this is most easy with text() where you'll get your exact SQL

One thing, I always get a bit confused by this part of SQLAlchemy.  How do I do this with Core? I can't meaningfully maintain a raw statement.

It's actually one of those things I wish was a tad easier.

**Reply by sonthonaxrk** [reply]
> OK aliased_class doesn't work here anyway because it can't locate the columns. this is because the "abstract base" is mapped to an arbitrary selectable, and in order to make any kind of adaptation on that, the new "selectable" has to be based on the old one.

I don't quite understand this.  This query - for reasons I don't know - contains a cartesian join and get 4 rows rather 2.  But it still maps correctly; which to me points that it was able to make an adaptation.


```
    result = session.execute(
        select(base_obj_alias).select_from(special_select)
    )
```

```
    res = result.all()
    assert len(res) == 4
```

Even this produces a cartesian join

```
```
    result = session.execute(
        select(base_obj_alias)
    )
```

**Reply by zzzeek** [reply]
> > is that a concern because...the JOIN that would happen between Metadata and a single class?
> 
> So effectively I have a series of well defined objects (A, B), but their primary key maps to the primary key of common metadata. And the primary API I've written is a polymorphic list of those objects. Because the ordering is usually (but not always) by the common metadata id, I have no guarantee what types will be returned by the query.
> 
> The other reason that load polymorphic on SELECT in won't work in my case is that I might want to filter each thing in the UNION query. Which is harder to do in in an optimal manner.

we can work out a way to make it filterable, idealy though this would just be the fixed polymorphic query for the base.    there's a method you should override on your AbstractConcreteBase called _create_polymoprhic_union():

```py
    @classmethod
    def _create_polymorphic_union(cls, mappers, discriminator_name):
        return polymorphic_union(
            OrderedDict(
                (mp.polymorphic_identity, mp.local_table) for mp in mappers
            ),
            discriminator_name,
            "pjoin",
        )

```

put your custom UNION there, then the BaseObject will be mapped to that selectable.  you wont have you use with_polymorphic on the outside and it should be mostly as flexible as the default abstract concrete base.

> 
> I'm under the impression the polymorphic join makes really wide queries, which is fine, but given I could get close to some arbitrary limit of table width, I'd rather use union.

joined inheritance only does that if you have it loading all the tables at once, which is not the default behavior for joined inheritance.   but then it means you have extra SELECT statements in order to get data from the other tables.

> 
> Also zzzeek, thanks for your help. Let me know if there's anything I can do for you.

theres really nothing :)   i get a lot of donations, which is great, but none of that can create more time for me, that's the main thing I dont have.

**Reply by sonthonaxrk** [reply]
I've actually already overriden `_create_polymoprhic_union` for a different reason: it doesn't work with composite types.

I'm afraid I'm loading all the tables at once.  The query I've written is actually super quick, it's just the mapping that's getting me down.

**Reply by sonthonaxrk** [reply]
One of the complications, and another reason I want to map this like so, is that I actually plan on passing filter expressions into my custom `polymorphic_union` function.

**Reply by zzzeek** [reply]
> I've actually already overriden `_create_polymoprhic_union` for a different reason: it doesn't work with composite types.
> 

but...that's the function that's already being used to generate the query that is used by default, which we would assume...works?   if not, then that's a different issue.

are you able to get a basic load to work correctly?  if so, then we can change the query to the one you want.

**Reply by sonthonaxrk** [reply]
Like, I get what you're describing is functionally equivalent, and I already have a working query that maps correctly; however because I'm loading all the tables at once, the UNION ALL query actually ends up being the most efficient way to filter and gives lighting fast performance.

This is one part of a really complex query nested inside a recursive CTE.  I'd rather get the mapper to work when the SQL is now tuned really nicely.

**Reply by sonthonaxrk** [reply]
> are you able to get a basic load to work correctly? if so, then we can change the query to the one you want.

Basic load works fine.  But one of the problem is that I need to alias the BaseObj query onto an arbitrary bit of Core SQL.  Like this

```
    special_union = polymorphic_union(
        dict((mp.polymorphic_identity, mp.local_table) for mp in [
            A.__mapper__,
            B.__mapper__,
        ]),
        "type",
        filter_clauses_to_apply_to_each_element=[Metadata.data == 'test']
    )

    base_trade_alias = with_polymorphic(
        BaseObj,
        [A, B],
        selectable=special_union,
        aliased=True,
    )

    res = select(base_trade_alias, Metadata).from_statement(special_union)
    # returns things where Metadata.data == test
```

I get that I could filter and join on the outer statement, and that's what I'm currently doing.  But the problem is that's ruinous for performance.  I've gotten a 1000x speed up like this.

**Reply by sonthonaxrk** [reply]
It would just be really good if I could map BaseObj onto an arbitrary statement and pull out joined objects.

**Reply by sonthonaxrk** [reply]
I think I might just give up and try something a bit dumb, like writing my own clause adapter and monkey patching the aliased class with it.

Doing mapping myself is just such a tedious thing.

**Reply by zzzeek** [reply]
Here's your original example again:

* with composite types on both classes
* using select(), not text()
* joins to Metadata inside each sub-select inside the Union
* using _create_polymorphic_union() so that you can select BaseObj directly.
* inserts a little bit of test data to ensure the round trip
* filters the query to demonstrate simple filtering / ordering

So there's no issue with this general idea, I think generalizing the polymorphic_union can get tedious, but the POC below shows the general thing can work.  Also concrete inheritance can generate very wide rows also if all your classes have different columns.

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import literal
from sqlalchemy import null
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy import union
from sqlalchemy.ext.declarative import AbstractConcreteBase
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.ext.declarative import declared_attr
from sqlalchemy.orm import composite
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import declared_attr
from sqlalchemy.orm import Session


Base = declarative_base()


class Metadata(Base):
    __tablename__ = "metadata"
    id = Column(
        Integer,
        primary_key=True,
        unique=True,
    )


class BaseObj(AbstractConcreteBase, Base):
    @declared_attr
    def id(cls):
        return Column(ForeignKey(Metadata.id), primary_key=True)

    @classmethod
    def _create_polymorphic_union(cls, mappers, discriminator_name):
        return union(
            select(
                A.id,
                A.thing1,
                A.x1,
                A.y1,
                null().label("thing2"),
                null().label("x2"),
                null().label("y2"),
                literal("a").label("type"),
            ).join(Metadata),
            select(
                B.id,
                null().label("thing1"),
                null().label("x1"),
                null().label("y1"),
                B.thing2,
                B.x2,
                B.y2,
                literal("b").label("type"),
            ).join(Metadata),
        ).subquery()


class XYThing:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __composite_values__(self):
        return (self.x, self.y)

    def __repr__(self):
        return f"XYThing({self.x}, {self.y})"


class A(BaseObj):
    __tablename__ = "a"
    thing1 = Column(String)
    comp1 = composite(XYThing, Column("x1", Integer), Column("y1", Integer))

    @declared_attr
    def __mapper_args__(cls):
        return {"polymorphic_identity": cls.__tablename__, "concrete": True}

    def __repr__(self):
        return f"A(id={self.id}, thing1={self.thing1}, comp1={self.comp1})"


class B(BaseObj):
    __tablename__ = "b"
    thing2 = Column(String)
    comp2 = composite(XYThing, Column("x2", Integer), Column("y2", Integer))

    @declared_attr
    def __mapper_args__(cls):
        return {"polymorphic_identity": cls.__tablename__, "concrete": True}

    def __repr__(self):
        return f"B(id={self.id}, thing2={self.thing2}, comp2={self.comp2})"


e = create_engine("sqlite://", echo='debug')
Base.metadata.create_all(e)

with Session(e) as sess:
    sess.add_all([Metadata(id=1), Metadata(id=2)])
    sess.flush()

    sess.add_all(
        [
            A(id=1, thing1="thing1", comp1=XYThing(1, 2)),
            B(id=2, thing2="thing2", comp2=XYThing(3, 4)),
        ]
    )
    sess.commit()


with Session(e) as sess:

    for obj in sess.scalars(select(BaseObj).where(BaseObj.id < 3).order_by(BaseObj.id)):
        print(obj)

```

**Reply by zzzeek** [reply]
the query emitted is:

```sql
SELECT anon_1.id, anon_1.thing1, anon_1.x1, anon_1.y1, anon_1.thing2, anon_1.x2, anon_1.y2, anon_1.type 
FROM (SELECT a.id AS id, a.thing1 AS thing1, a.x1 AS x1, a.y1 AS y1, NULL AS thing2, NULL AS x2, NULL AS y2, ? AS type 
FROM a JOIN metadata ON metadata.id = a.id UNION SELECT b.id AS id, NULL AS thing1, NULL AS x1, NULL AS y1, b.thing2 AS thing2, b.x2 AS x2, b.y2 AS y2, ? AS type 
FROM b JOIN metadata ON metadata.id = b.id) AS anon_1 
WHERE anon_1.id < ? ORDER BY anon_1.id
2022-03-07 19:45:18,296 INFO sqlalchemy.engine.Engine [generated in 0.00031s] ('a', 'b', 3)
2022-03-07 19:45:18,297 DEBUG sqlalchemy.engine.Engine Col ('id', 'thing1', 'x1', 'y1', 'thing2', 'x2', 'y2', 'type')
2022-03-07 19:45:18,297 DEBUG sqlalchemy.engine.Engine Row (1, 'thing1', 1, 2, None, None, None, 'a')
2022-03-07 19:45:18,297 DEBUG sqlalchemy.engine.Engine Row (2, None, None, None, 'thing2', 3, 4, 'b')
A(id=1, thing1=thing1, comp1=XYThing(1, 2))
B(id=2, thing2=thing2, comp2=XYThing(3, 4))

```

**Reply by zzzeek** [reply]
also the composite types work with the default query as well

**Reply by sonthonaxrk** [reply]
Ahh, man I'm so sorry.  The problem is (and the reason I was trying to map a custom union onto BaseObb) was that I needed to apply a filter condition, dynamically, to each and every one of the UNION queries.

It would ultimately look something like this

```    @classmethod
    def _create_polymorphic_union(cls, mappers, discriminator_name):
        return union(
            select(
                A.id,
                A.thing1,
                A.x1,
                A.y1,
                null().label("thing2"),
                null().label("x2"),
                null().label("y2"),
                literal("a").label("type"),
            ).join(Metadata).where(*filter_clauses),
            select(
                B.id,
                null().label("thing1"),
                null().label("x1"),
                null().label("y1"),
                B.thing2,
                B.x2,
                B.y2,
                literal("b").label("type"),
            ).join(Metadata).where(*filter_clauses),
        ).subquery()
```

I get you're suggesting to filter the union, but it just kills performance massively, and to get decent performance, I need to filter each and every part of the union query.  Like it's non-trivial performance increase (1000x) from something that's semantically equivalent. 

That's kinda why I wanted to map onto an arbitrary statement.

I'll try editing what you've posted to see if I can get it to work.

**Reply by sonthonaxrk** [reply]
Also, again thanks for all your help.  I appreciate I'm being a bit difficult here, and I've got a weird problem.

**Reply by sonthonaxrk** [reply]
Could I create new BaseObjs for each query I want to run?  I appreciate that's kinda like an aliased class and is super dirty.

**Reply by sonthonaxrk** [reply]
> also the composite types work with the default query as well

Indeed it does with SQLite, but not with Postgres.  But it's such a fringe issue (and easily fixed) that it's not worth bothering you about.

**Reply by sonthonaxrk** [reply]
It's just an unfortunate fact that I have to dynamically generate the UNION query.  It's not that what you're suggesting is wrong, it is semantically the same, however Postgres won't see the WHERE clause as distributive over each clause in the UNION query, so instead each part of the UNION query will actually load up the entire table in memory.  So I have to filter each part of the UNION query manually.

You're totally right about what would be simple to do, but this performance problem is why I'm trying to map BaseObj dynamically against an arbitrary select at runtime:  

```python
with Session(e) as sess:

    # I get this is semantically the same as what you suggested.
    # But performance wise, is incomparably faster.
    clauses = [Metadata.id > 3]

    my_union = union(
        select(
            A.id,
            A.thing1,
            A.x1,
            A.y1,
            null().label("thing2"),
            null().label("x2"),
            null().label("y2"),
            literal("a").label("type"),
        ).join(Metadata).filter(*clauses),
        select(
            B.id,
            null().label("thing1"),
            null().label("x1"),
            null().label("y1"),
            B.thing2,
            B.x2,
            B.y2,
            literal("b").label("type"),
        ).join(Metadata).filter(*clauses),
    ).subquery()

    base_obj_alias = with_polymorphic(
        BaseObj,
        [A, B],
        selectable=my_union,
        aliased=True,
        _use_mapper_path=True
    )

    for obj in sess.scalars(select(base_obj_alias)):
        print(obj)
        # (A(id=1, thing1=thing1, comp1=XYThing(1, 2)),)
        # (B(id=1, thing2=None, comp2=XYThing(None, None)),)
        # (A(id=2, thing1=None, comp1=XYThing(None, None)),)
        # (B(id=2, thing2=thing2, comp2=XYThing(3, 4)),)

```


However, this isn't working _because it keeps on making a cartesian join_.  It does this: `SELECT my_union_1, my_union_2 ...`

**Reply by sonthonaxrk** [reply]
I feel like this cartesian join issue should be simple enough to solve.  However I think I should sleep on it before I start digging into the compiler to work out why it adds two entities

**Reply by sonthonaxrk** [reply]
The other thing is also that I need to get the Metadata object out of this too.  But again, I just can't map it.  I can't use contains_eager, and adding it to the select entities just creates yet another Cartesian join

**Reply by zzzeek** [reply]
There's no problem with PostgreSQL on this end. Also the main thing that's needed to map arbitrary selectables at query time is a means of matching up the columns with the mapped statement.   This can be done based on name matching as follows:

```py
def make_statement(*filter_cond):

    return union(
        select(
            A.id,
            A.thing1,
            A.x1,
            A.y1,
            null().label("thing2"),
            null().label("x2"),
            null().label("y2"),
            literal("a").label("type"),
        )
        .join(Metadata)
        .filter(*filter_cond),
        select(
            B.id,
            null().label("thing1"),
            null().label("x1"),
            null().label("y1"),
            B.thing2,
            B.x2,
            B.y2,
            literal("b").label("type"),
        )
        .join(Metadata)
        .filter(*filter_cond),
    )

with Session(e) as sess:

    ac = aliased(
        BaseObj,
        alias=make_statement(Metadata.id < 3).subquery(),
        adapt_on_names=True,
    )
    for obj in sess.scalars(select(ac)):
        print(obj)
```



> The other thing is also that I need to get the Metadata object out of this too. But again, I just can't map it. I can't use contains_eager, and adding it to the select entities just creates yet another Cartesian join

I dont think joined eager loading works with AbstractConcreteBase in any case, at least I can get it to make the whole query with contains_eager() but the mechanics dont quite know how to fetch the related object for both A and B individually so it does not actually work.    However, you can use selectinload(), which is extremely efficient.   It also works if you select BaseObj and Metadata in one query but just not as a related item to the base, i will include that in the demo below.

demo below works on PostgreSQL, adds filter criteria to each sub-select, and uses selectinload to get related metadata:

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import literal
from sqlalchemy import null
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy import union
from sqlalchemy.ext.declarative import AbstractConcreteBase
from sqlalchemy.orm import aliased
from sqlalchemy.orm import composite
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import declared_attr
from sqlalchemy.orm import relationship
from sqlalchemy.orm import selectinload
from sqlalchemy.orm import Session


Base = declarative_base()


class Metadata(Base):
    __tablename__ = "metadata"
    id = Column(
        Integer,
        primary_key=True,
        unique=True,
    )


def make_statement(*filter_cond, include_metadata=False):

    a_stmt = (
        select(
            A.id,
            A.thing1,
            A.x1,
            A.y1,
            null().label("thing2"),
            null().label("x2"),
            null().label("y2"),
            literal("a").label("type"),
        )
        .join(Metadata)
        .filter(*filter_cond)
    )
    if include_metadata:
        a_stmt = a_stmt.add_columns(Metadata.__table__.c.id.label("metadata_id"))

    b_stmt = (
        select(
            B.id,
            null().label("thing1"),
            null().label("x1"),
            null().label("y1"),
            B.thing2,
            B.x2,
            B.y2,
            literal("b").label("type"),
        )
        .join(Metadata)
        .filter(*filter_cond)
    )
    if include_metadata:
        b_stmt = b_stmt.add_columns(Metadata.__table__)

    return union(a_stmt, b_stmt)


class BaseObj(AbstractConcreteBase, Base):
    @declared_attr
    def id(cls):
        return Column(ForeignKey(Metadata.id), primary_key=True)

    @classmethod
    def _create_polymorphic_union(cls, mappers, discriminator_name):
        return make_statement().subquery()

    @declared_attr
    def related_metadata(cls):
        return relationship(Metadata)


class XYThing:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __composite_values__(self):
        return (self.x, self.y)

    def __repr__(self):
        return f"XYThing({self.x}, {self.y})"


class A(BaseObj):
    __tablename__ = "a"
    thing1 = Column(String)
    comp1 = composite(XYThing, Column("x1", Integer), Column("y1", Integer))

    @declared_attr
    def __mapper_args__(cls):
        return {"polymorphic_identity": cls.__tablename__, "concrete": True}

    def __repr__(self):
        return f"A(id={self.id}, thing1={self.thing1}, comp1={self.comp1})"


class B(BaseObj):
    __tablename__ = "b"
    thing2 = Column(String)
    comp2 = composite(XYThing, Column("x2", Integer), Column("y2", Integer))

    @declared_attr
    def __mapper_args__(cls):
        return {"polymorphic_identity": cls.__tablename__, "concrete": True}

    def __repr__(self):
        return f"B(id={self.id}, thing2={self.thing2}, comp2={self.comp2})"


e = create_engine("postgresql://scott:tiger@localhost/test", echo="debug")
Base.metadata.drop_all(e)
Base.metadata.create_all(e)

with Session(e) as sess:
    sess.add_all([Metadata(id=1), Metadata(id=2)])
    sess.flush()

    sess.add_all(
        [
            A(id=1, thing1="thing1", comp1=XYThing(1, 2)),
            B(id=2, thing2="thing2", comp2=XYThing(3, 4)),
        ]
    )
    sess.commit()


# version one - use selectinload

with Session(e) as sess:

    alias = make_statement(Metadata.id < 3).subquery()
    ac = aliased(
        BaseObj,
        alias=alias,
        adapt_on_names=True,
    )

    # will emit two queries, one is the UNION, the second a simple
    # SELECT for the Metadata objects
    for obj in sess.scalars(
        select(ac).options(selectinload(ac.related_metadata))
    ):

        print(obj)
        print(obj.related_metadata)

# version two - we can load Metadata at the same time just not
# as "related_metadata"

with Session(e) as sess:

    alias = make_statement(Metadata.id < 3, include_metadata=True).subquery()
    ac = aliased(
        BaseObj,
        alias=alias,
        adapt_on_names=True,
    )
    ma = aliased(Metadata, alias=alias)

    # will load both objects at the same time
    for obj, md_obj in sess.execute(
        select(ac, ma)
    ):

        print(obj)
        print(md_obj)

```

**Reply by zzzeek** [reply]
OK I can add support, somewhat awkward at the moment but it works, for the contains_eager part in the next 1.4 release with two small changes, so you'd get the whole query w contains eager like this:

```py
with Session(e) as sess:

    alias = make_statement(Metadata.id < 3, include_metadata=True).subquery()
    ac = with_polymorphic(
        BaseObj,
        [A, B],
        selectable=alias,
        adapt_on_names=True,
    )

    mt = aliased(Metadata, alias=alias)

    for obj in sess.scalars(
        select(ac).options(
            contains_eager(ac.A.related_metadata.of_type(mt)),
            contains_eager(ac.B.related_metadata.of_type(mt)),
        )
    ):

        print(obj)
        print(obj.related_metadata)

```


I dont know if you actually want to use this, since this is way out there, but I'm just slowly pushing little things here and there to make these cases work better each time.
